lib/state_change_syncer.go

```diff
@@ -0,0 +1,726 @@
+}
+
+// Add a transaction to the queue of transactions to be flushed to disk upon badger db flush.
+func (stateChangeSyncer *StateChangeSyncer) addTransactionToQueue(flushId uuid.UUID, writeBytes []byte) {
+	stateChangeSyncer.EntryCount++
+
+	unflushedBytes, exists := stateChangeSyncer.UnflushedBytes[flushId]
+	if !exists {
+		unflushedBytes = UnflushedStateSyncerBytes{
+			StateChangeBytes:            []byte{},
+			StateChangeBytesSize:        0,
+			StateChangeOperationIndexes: []uint64{},
+		}
+	}
+
+	unflushedBytes.StateChangeBytes = append(unflushedBytes.StateChangeBytes, writeBytes...)
+
+	// Get the byte index of where this transaction occurs in the unflushed bytes, and add it to the list of
+	// indexes that should be written to the index file.
+	dbOperationIndex := unflushedBytes.StateChangeBytesSize
+	unflushedBytes.StateChangeOperationIndexes = append(unflushedBytes.StateChangeOperationIndexes, dbOperationIndex)
+
+	// Update the state change file size.
+	transactionLen := uint64(len(writeBytes))
+	unflushedBytes.StateChangeBytesSize += transactionLen
+
+	stateChangeSyncer.UnflushedBytes[flushId] = unflushedBytes
+}
+
+// FlushTransactionsToFile writes the bytes that have been cached on the StateChangeSyncer to the state change file.
+func (stateChangeSyncer *StateChangeSyncer) FlushTransactionsToFile(event *StateSyncerFlushedEvent) error {
+	flushId := event.FlushId
+	// Get the relevant global flush ID from the state change syncer if the flush ID is nil.
+	if event.IsMempoolFlush && event.FlushId == uuid.Nil {
+		flushId = stateChangeSyncer.MempoolFlushId
+	} else if !event.IsMempoolFlush && event.FlushId == uuid.Nil {
+		flushId = stateChangeSyncer.BlockSyncFlushId
+	}
+
+	var flushFile *os.File
+	var flushFileSize uint64
+	var indexFile *os.File
+	if event.IsMempoolFlush {
+		flushFile = stateChangeSyncer.StateChangeMempoolFile
+		flushFileSize = stateChangeSyncer.StateChangeMempoolFileSize
+		indexFile = stateChangeSyncer.StateChangeMempoolIndexFile
+	} else {
+		flushFile = stateChangeSyncer.StateChangeFile
+		flushFileSize = stateChangeSyncer.StateChangeFileSize
+		indexFile = stateChangeSyncer.StateChangeIndexFile
+	}
+
+	// If the flush failed, delete the unflushed bytes and associated metadata.
+	// Also delete any unconnected mempool txns from our cache.
+	if !event.Succeeded {
+		delete(stateChangeSyncer.UnflushedBytes, flushId)
+	}
+
+	unflushedBytes, exists := stateChangeSyncer.UnflushedBytes[flushId]
+
+	if !exists {
+		return nil
+	}
+
+	if unflushedBytes.StateChangeBytes == nil || unflushedBytes.StateChangeOperationIndexes == nil || len(unflushedBytes.StateChangeOperationIndexes) == 0 {
+		return fmt.Errorf("Error flushing state change file: FlushId %v has nil bytes\n", flushId)
+	}
+
+	// Write the encoded StateChangeEntry bytes to the state changer file.
+	_, err := flushFile.Write(unflushedBytes.StateChangeBytes)
+
+	if err != nil {
+		return fmt.Errorf("Error writing to state change file: %v", err)
```

Comment:
May be useful to distinguish whether this failure happened on a mempool flush or a main db one. Latter is pretty serious so we'd want to see that in the logs.